lib/collection/src/hash_ring.rs

```diff
@@ -22,48 +22,18 @@ pub enum HashRingRouter<T = ShardId> {
     Resharding { old: HashRing<T>, new: HashRing<T> },
 }
 
-impl<T: Hash + Copy + PartialEq> HashRingRouter<T> {
+impl<T: Copy + Eq + Hash> HashRingRouter<T> {
     /// Create a new single hashring.
     ///
     /// The hashring is created with a fair distribution of points and `HASH_RING_SHARD_SCALE` scale.
     pub fn single() -> Self {
         Self::Single(HashRing::fair(HASH_RING_SHARD_SCALE))
     }
 
-    /// Create a new resharding hashring, with resharding shard already added into `new` hashring.
-    ///
-    /// The hashring is created with a fair distribution of points and `HASH_RING_SHARD_SCALE` scale.
-    pub fn resharding(shard: T, direction: ReshardingDirection) -> Self {
-        let mut ring = Self::Resharding {
-            old: HashRing::fair(HASH_RING_SHARD_SCALE),
-            new: HashRing::fair(HASH_RING_SHARD_SCALE),
-        };
-
-        ring.start_resharding(shard, direction);
-
-        ring
-    }
-
-    pub fn is_empty(&self) -> bool {
-        match self {
-            Self::Single(ring) => ring.is_empty(),
-            Self::Resharding { old, new } => old.is_empty() && new.is_empty(),
-        }
-    }
-
-    pub fn is_resharding(&self) -> bool {
-        matches!(self, Self::Resharding { .. })
-    }
-
-    pub fn add(&mut self, shard: T) {
+    pub fn add(&mut self, shard: T) -> bool {
         match self {
             Self::Single(ring) => ring.add(shard),
-            Self::Resharding { old, new } => {
-                if new.get(&shard).is_none() {
-                    old.add(shard);
-                    new.add(shard);
-                }
-            }
+            Self::Resharding { .. } => panic!("can't add shards to hashring in resharding mode"),
```

Comment:
Not sure I like this hard panic here.

The reason is that it might be very annoying to resolve in a customers' cluster if it ends up in WAL.

I'd prefer to do a debug assert instead, and do some best-effort approach in release builds.

Wdyt?

Reply:
I don't think it can ever end up in WAL, though.

Reply:
I hit this panic right away during testing, when resharding is started through consensus.

Reply:
We agreed to revert this to initial behavior of `add`, because it requires a more thorough refactoring, which we can introduce later.